proposals/nnnn-trailing-comma-lists.md

```diff
@@ -0,0 +1,232 @@
+# Allow trailing comma in comma-separated lists
+
+- Proposal: SE-NNNN
+- Author: Mateus Rodrigues
+- Review Manager: TBD
+- Status: Awaiting Review
+- Implementation: https://github.com/swiftlang/swift/pull/74522# gated behind `-enable-experimental-feature TrailingComma`
+- Review: [pitch](https://forums.swift.org/t/pitch-allow-trailing-comma-in-tuples-arguments-and-if-guard-while-conditions/70170/48)
+
+## Introduction
+
+This proposal aims to allow the use of trailing commas, currently restricted to array and dictionary literals, in comma-separated lists whenever there are terminators that enable unambiguous parsing.
+
+## Motivation
+
+### Development Quality of Life Improvement
+
+A trailing comma is an optional comma after the last item in a list of elements:
+
+```swift
+let rank = [
+  "Player 1",
+  "Player 3",
+  "Player 2",
+]
+```
+
+Swift's support for trailing commas in array and dictionary literals makes it as easy to append, remove, reorder, or comment out the last element as any other element.
+
+Other comma-separated lists in the language could also benefit from the flexibility enabled by trailing commas. Consider the [split(separator:maxSplits:omittingEmptySubsequences:)](https://swiftpackageindex.com/apple/swift-algorithms/1.2.0/documentation/algorithms/swift/lazysequenceprotocol/split(separator:maxsplits:omittingemptysubsequences:)-4q4x8) function from the [Algorithms](https://github.com/apple/swift-algorithms) package, which has a few overloads due to default values.
+
+`split(separator)` and `split(separator:maxSplits)` are both valid calls but you can't easily switch between them without adding/removing the trailing comma.
+
+```swift
+let numbers = [1, 2, 0, 3, 4, 0, 0, 5]
+    
+let subsequences = numbers.split(
+	separator: 0,
+//        maxSplits: 1
+) ❌ Unexpected ',' separator
+```
+
+### The Language Evolved
+
+Back in 2016, a similar [proposal](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0084-trailing-commas.md) with a narrower scope was reviewed and rejected for Swift 3 but since that time the language has evolved substantially that changed the basis for rejection. The code style that "puts the terminating right parenthesis on a line following the arguments to that call" has been widely adopted by community, swift standard library codebase, swift-format, docc documentation and Xcode. Therefore, not encouraging or endorsing this code style doesn't hold true anymore.
+The language  has also seen the introduction of [parameter packs](https://github.com/apple/swift-evolution/blob/main/proposals/0393-parameter-packs.md), which allows more APIs that are comma-separated lists at call site and would benefit from trailing comma, and code generation tools like plugins and macros that with trailing comma support wouldn't have to worry about a special condition for the last element when generating comma-separated lists.
```

Comment:
```suggestion
The language has also seen the introduction of [parameter packs](https://github.com/apple/swift-evolution/blob/main/proposals/0393-parameter-packs.md), which enables APIs that are generic over variable numbers of type parameters, and code generation tools like plugins and macros that, with trailing comma support, wouldn't have to worry about a special condition for the last element when generating comma-separated lists.
```

Any function that takes more than one argument would have a comma-separated argument list; the salient point is that now these APIs can have _variable_ numbers of type parameters.